Networking

Stiffly opened the pull request:
### Edit:

Network logic not using threads anymore
Renamed Package -> Packet
Renamed PopFrontData() -> ReadData(). AddData() -> WriteData().
Renamed PrivateMemberFunctions() -> privateMemberFunctions()
Utilized configfile for address, port, player's name, chosing between server and client.
Deleted NetworkDefinitions.h and restructured its members.
Fixed lines in .ini files.

<b>PlayerSystem</b>
PlayerSystem does not read any input anymore, now updates movement. (with temporary implemented player component)
<b>Rectangle class</b>
Engines Rectangle class conflicted with Asio's rectangle class in Game.cpp for instance. Now using Rectangle::Rectangle(...) to refer to the correct implementation.
<b>Package class</b>
Implemented a Package class to abstract the message passing process with:
### Package.h

``` h
class Package {
    AddPrimitive<type>() { ... } // Adds a primitive type to a package
    AddString();
    AddData();
    PopFrontPrimitive<type>() { ... } // Reads a primitive type from a package
    PopFrontString();
    PopFrontData();
}
```
### Example:

``` cpp
Package package(MessageType::Connect, m_PacketID);
package.AddPrimitive<int>(5); // Player ID
int playerID = package.PopFrontPrimitive<int>(); // playerID will get value of 5
```

<b>Logic:</b>
The network runs on a different thread than the rest of game but has an update function where critical logic can be placed as this will run on the game thread .
Client sends input to server.
Server does movement logic and return transform component.
Client updates its own movement but is forced to the transform component received from the server.

<b>How to use:</b>
Make sure your Networking variable in Config.ini is set to true.
In console write 'C' and press enter for Client and 'S' for Server.
When a server is running and you want to connect with a Client, press 'C' in game.

Stiffly committed:
Added utility headers for network interface.

Stiffly committed:
Rectangle definition conflicted with a definition included by <boost/asio.hpp>.

Stiffly committed:
Copy pasta error. Corrected #ifndef

Stiffly committed:
Added the old prototype logic to Client.h. Can now communicate with a external server(on prototype).

Stiffly committed:
Merge remote-tracking branch 'origin/master' into Networking

JockeBTH committed:
Network now runs on it's own thread.

Network now sends Disconnect message when game is terminated.

JockeBTH committed:
Merge branch 'Networking' of github.com:teamfisk/TacticalZ into Networking

# Conflicts:
#	include/Game/Game.h

JockeBTH committed:
Added leak check when using windows (WinLeakCheck.h).

Added server to project.
A server can now be started in the project.

Stiffly committed:
WIP trying to get Eventsystem to work in client.

Stiffly committed:
Merge branch 'Networking' of github.com:teamfisk/TacticalZ into Networking

# Conflicts:
#	include/Engine/Network/Client.h
#	src/Game/Game.cpp

JockeBTH committed:
Added support for multiplayer.

Reactored code, clumped together variables in PlayerDefinition.h.
Integrated server with entity system.

Stiffly committed:
Input events for Client done. Instead of windows getAsyncKeyState();

JockeBTH committed:
Merge branch 'Networking' of github.com:teamfisk/TacticalZ into Networking

JockeBTH committed:
Cleaned up code in Client.h and Client.cpp.

Moved ping code into Ping() function.

Stiffly committed:
Implemented new version of snapshot for 3D world. Also some cleanup.

Stiffly committed:
Merge branch 'Networking' of github.com:teamfisk/TacticalZ into Networking

# Conflicts:
#	include/Engine/Network/Client.h
#	src/Engine/Network/Client.cpp

Stiffly committed:
Random bugfixes and cleanup.

Stiffly committed:
Bugfix. if-statement was incorrect

Stiffly committed:
Added a "StartNetwork" bool to DefaultConfig.

Stiffly committed:
Renamed NetworkDefines.h --> NetworkDefinitions.h

Stiffly committed:
Smoother movement over network. Some refactoring on client side. Now uses EKeyUp to determine if a key is pressed.

JockeBTH committed:
Fixed bug where closing glwindow as client crashed.

Server bug is still present.

JockeBTH committed:
Fixed bug that cause the program to crash when closing if a server had been started.

Removed memory leaks in Client.cpp.

JockeBTH committed:
Fixed the bug where starting a server made the program crash on closing.

Fixed some memory leaks in Client.cpp.

Stiffly committed:
WIP Packet loss identification. Put packetID in message

JockeBTH committed:
Moved client and server from header.

Fixed so user are able to start a server and client on the same computer.
Starting 2 servers on the same computer will probably still crash the newest server (Don't do it).

Stiffly committed:
Can now Identify Packet Loss. Enumerated messages that loops around 0-1000.

Stiffly committed:
Merge remote-tracking branch 'origin/master' into Networking

NOTE (in config file):
[Debug]
LoadMap=Schema/Entities/Test.xml

# Conflicts:
#	include/Game/Game.h
#	src/Game/Game.cpp

JockeBTH committed:
Fixed client movement.

JockeBTH committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	include/Engine/Network/NetworkDefinitions.h
#	include/Game/Game.h
#	src/Engine/Network/Client.cpp
#	src/Game/Game.cpp

Stiffly committed:
Added class Package to help refactor the code. The code got easier to read.

Stiffly committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	src/Engine/Network/Client.cpp

Stiffly committed:
Merge branch 'master' into Networking

JockeBTH committed:
Reactored the package loss logic.

Server cannot handle multiple clients yet.
Fixed WorldTest code by adding astrix.

JockeBTH committed:
Merge branch 'Networking' of github.com:teamfisk/TacticalZ into Networking

Stiffly committed:
Merge branch 'PlayerSystem' into Networking

# Conflicts:
#	src/Game/Game.cpp

Stiffly committed:
Indentation

JockeBTH committed:
Fixed bug which made the player not able to move diagonally.

Fixed a bug were m_PlayerID was not set for the client when connecting.

Stiffly committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	src/Engine/Network/Client.cpp

JockeBTH committed:
IdentifyPacketLoss() crashed when multiple users connected which

also crashed the server. The function is not needed for now as the server logic is not
implemented.

Stiffly committed:
Merge remote-tracking branch 'origin/Networking' into Networking

Stiffly committed:
Removed annoying cube

Stiffly committed:
Merge remote-tracking branch 'origin/master' into Networking

# Conflicts:
#	include/Game/Game.h
#	resources/DefaultConfig.ini
#	src/Game/Game.cpp

Stiffly committed:
WIP

Adding models now crashes. Threading
Added a create player event that currently is not used.

Stiffly committed:
Merge remote-tracking branch 'origin/master' into Networking

# Conflicts:
#	src/Engine/Rendering/RenderState.cpp
#	src/Game/Game.cpp

Stiffly committed:
Refactoring. Client and server has Update Loop that runs before any game logic each frame.

JockeBTH committed:
WIP

Stiffly committed:
Various bug fixes.

Stiffly committed:
Fixed bug with entityID. Also trying to fix movement WIP.

Stiffly committed:
Cleaned up PlayerSystem. Sets velocity to 0 every frame.

Stiffly committed:
Updated movement code once again.

Stiffly committed:
Removed old input events from client

Stiffly committed:
Test code

JockeBTH committed:
Made it easier for non network members to create packages and extract information from them by addin PopFrontString() and PopFrontPrimitive<type>().

Stiffly committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	src/Engine/Network/Client.cpp
#	src/Engine/Network/Server.cpp

Stiffly committed:
Movement code not buggy anymore. Also changed package to not allocate more data than needed.

JockeBTH committed:
Added a class that client and server inherits from.

Class and server now share the same pointer i game.
You are now able to start a client and a server on the same computer.

JockeBTH committed:
Merge remote-tracking branch 'origin/Networking' into Networking

JockeBTH committed:
Restructured code and added bool to stop client from spamming Snapshots to a server when

it wasn't connected to one.

JockeBTH committed:
Fixed "trying to read more than package size" warnings.

Stiffly committed:
Can now send transform components.

JockeBTH committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	src/Engine/Network/Client.cpp

Stiffly committed:
Cleaned code

Stiffly committed:
WIP Cleaning

Stiffly committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	include/Engine/Network/Client.h
#	src/Engine/Network/Client.cpp

Stiffly committed:
More cleaning code

Stiffly committed:
Merge remote-tracking branch 'origin/master' into Networking

# Conflicts:
#	src/Game/Game.cpp

Stiffly committed:
No more fake packet loss

Stiffly committed:
Removed old Event that is not used

Stiffly committed:
Reset RenderQueueFactory.cpp to original state

sippeangelo commented on include/Engine/Network/NetworkDefinitions.h:
```diff
@@ -0,0 +1,20 @@
+#include "Network/Package.h"
+
+
+#define BOARDSIZE 16
```
wtf is this

sippeangelo commented on include/Engine/Network/NetworkDefinitions.h:
```diff
@@ -0,0 +1,20 @@
+#define BOARDSIZE 16
+#define MAXCONNECTIONS 8
+#define INPUTSIZE 128
+#define PLAYERSPEED 0.2f;
```
wtf is this

sippeangelo commented on include/Engine/Network/Package.h:
```diff
@@ -0,0 +1,56 @@
+public:
+    // arg1: Type of message (Connect, Disconnect...)
+    // arg2: PackageID for identifying packet loss.
+    Package(MessageType type, unsigned int& packageID);
```
Packet != Package

sippeangelo commented on src/Engine/Network/Package.cpp:
```diff
@@ -0,0 +1,62 @@
+    // Add message type
+    int messageType = static_cast<int>(type);
+    Package::AddPrimitive<int>(messageType);
+    packageID = packageID % 1000; // Packet id modulos
```
?

sippeangelo commented on src/Engine/Network/Package.cpp:
```diff
@@ -0,0 +1,62 @@
+
+Package::Package(MessageType type, unsigned int& packageID)
+{
+    m_Data = new char[128];
```
Template this up and base size on message type?

sippeangelo commented on src/Engine/Network/Package.cpp:
```diff
@@ -0,0 +1,62 @@
+void Package::AddString(std::string str)
+{
+    // Message, add one extra byte for null terminator
+    memcpy(m_Data + m_Offset, str.data(), (str.size() + 1) * sizeof(char));
```
Missing null terminator?

Stiffly replied:
Probably not

sippeangelo commented on src/Engine/Network/Package.cpp:
```diff
@@ -0,0 +1,62 @@
+{
+    std::string returnValue(m_Data + m_ReturnDataOffset);
+    if (m_Offset < m_ReturnDataOffset + returnValue.size()){
+        LOG_WARNING("Package PopFrontString(): Oh no! You are trying to remove things outside my memory kingdom");
```
lol

sippeangelo commented on resources/DefaultConfig.ini:
```diff
@@ -1,11 +1,19 @@
 [Debug]
+
```
Fix line endings

sippeangelo commented on src/Engine/Input/InputProxy.cpp:
```diff
@@ -62,7 +62,7 @@ void InputProxy::Process()
             e.Value = currentValue;
             m_EventBroker->Publish(e);
-            LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
+            //LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
```
Revert

sippeangelo commented on src/Engine/Input/InputProxy.cpp:
```diff
@@ -78,7 +78,7 @@ void InputProxy::Process()
         //e.Value = std::max(-1.f, std::min(e.Value, 1.f));
         m_EventBroker->Publish(e);
-        LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
+        //LOG_DEBUG("Input: Published command %s=%f for player %i", e.Command.c_str(), e.Value, e.PlayerID);
```
Revert

sippeangelo commented on src/Engine/Network/Client.cpp:
```diff
@@ -1,11 +1,347 @@
 
+Client::Client() : m_Socket(m_IOService)
+{
+    m_ReceiverEndpoint = udp::endpoint(boost::asio::ip::address::from_string("192.168.1.2"), 13);
```
*cough*

sippeangelo commented on src/Engine/Network/Client.cpp:
```diff
@@ -1,11 +1,347 @@
 Client::~Client()
 {
+    // will it work on linux? #if defined (_WIN64) | defined(_WIN32) otherwise.
+    _CrtDumpMemoryLeaks();
```
NO

sippeangelo commented on src/Game/Game.cpp:
```diff
@@ -53,7 +53,9 @@ Game::Game(int argc, char* argv[])
     m_SystemPipeline->AddSystem<EditorSystem>(m_Renderer);
-
+    // Invoke network
+    if (m_Config->Get<bool>("Networking.StartNetwork", false) == true)
```
https://github.com/teamfisk/TacticalZ/blob/master/tools/CodeRules.h

Stiffly committed:
Fixes, mostly name changes

JockeBTH committed:
Added more error code to Packet class

Stiffly committed:
Removed WinLeakCheck. Minor variable name changes. removed IsWASDKeyDown struct

JockeBTH committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	include/Engine/Network/Packet.h
#	src/Engine/Network/Packet.cpp

JockeBTH committed:
Made Network.h's Start() and Update pure virtual and removed cpp.

Removed unnecessary code from NetworkDefinitions.h.
Renamed sizeOfPackage to sizeOfPacket in Packet.cpp.

JockeBTH committed:
Removed NetworkDefinitions.h and moved #define MAXCONNECTIONS 8

#define INPUTSIZE 128 to Network.h.

JockeBTH committed:
Removed magic number from Packet constructor in Packet.cpp.

JockeBTH committed:
Merge remote-tracking branch 'origin/Networking' into Networking

Stiffly committed:
Client now reads address and port from config file. Default address is set to local host.

Stiffly committed:
Merge remote-tracking branch 'origin/Networking' into Networking

# Conflicts:
#	src/Engine/Network/Client.cpp

Stiffly committed:
Network not longer threaded

Stiffly committed:
Now uses config file to determine if client or server should be started.

Stiffly committed:
Merge remote-tracking branch 'origin/master' into Networking

# Conflicts:
#	include/Game/PlayerSystem.h
#	src/Game/Game.cpp
#	src/Game/PlayerSystem.cpp

Stiffly committed:
Added error check in Collision.cpp

Stiffly committed:
Re-enabled InputCommand debugging info in console
